File: c4/c2/persona.py
"""
c2 persona learning module.

Analyzes diffs between AI drafts and user final edits to extract
patterns, then proposes profile updates.
"""

from __future__ import annotations

import difflib
from pathlib import Path

from c4.c2.models import EditPattern, ProfileDiff


class PersonaLearner:
    """Learns user preferences from edit patterns across all four domains."""

    @staticmethod
    def analyze_edits(original: str, edited: str) -> list[EditPattern]:
        """Extract edit patterns from AI draft vs user final version.

        Args:
            original: AI-generated draft text.
            edited: User's final edited text.

        Returns:
            List of EditPattern objects describing the changes.
        """
        patterns: list[EditPattern] = []

        orig_lines = original.splitlines(keepends=True)
        edit_lines = edited.splitlines(keepends=True)

        diff = list(difflib.unified_diff(orig_lines, edit_lines, lineterm=""))

        # Categorize changes
        deletions: list[str] = []
        additions: list[str] = []

        for line in diff:
            if line.startswith("---") or line.startswith("+++") or line.startswith("@@"):
                continue
            if line.startswith("-"):
                deletions.append(line[1:].strip())
            elif line.startswith("+"):
                additions.append(line[1:].strip())

        # Detect tone changes (softening language)
        tone_softening = _detect_tone_softening(deletions, additions)
        if tone_softening:
            patterns.append(
                EditPattern(
                    category="tone",
                    description="User softened tone: assertive → questioning/hedging",
                    examples=tone_softening[:3],
                )
            )

        # Detect conciseness (user shortened text)
        if len(edited) < len(original) * 0.8:
            ratio = len(edited) / max(len(original), 1)
            patterns.append(
                EditPattern(
                    category="structure",
                    description=f"User significantly shortened text (ratio: {ratio:.2f})",
                    examples=[f"Original: {len(original)} chars → Edited: {len(edited)} chars"],
                )
            )

        # Detect section restructuring
        if _detect_section_reorder(original, edited):
            patterns.append(
                EditPattern(
                    category="structure",
                    description="User reordered sections",
                )
            )

        # Detect deletion patterns (what user removes)
        if deletions and not additions:
            patterns.append(
                EditPattern(
                    category="deletion",
                    description="User removed content without replacement",
                    examples=deletions[:3],
                )
            )

        # Detect addition patterns (what user adds)
        if additions and not deletions:
            patterns.append(
                EditPattern(
                    category="addition",
                    description="User added new content",
                    examples=additions[:3],
                )
            )

        # Detect wording substitutions
        substitutions = _detect_substitutions(deletions, additions)
        if substitutions:
            patterns.append(
                EditPattern(
                    category="wording",
                    description="User substituted specific wordings",
                    examples=substitutions[:3],
                )
            )

        return patterns

    @staticmethod
    def suggest_profile_updates(
        patterns: list[EditPattern],
    ) -> ProfileDiff:
        """Propose profile updates based on accumulated patterns.

        Args:
            patterns: List of edit patterns from multiple sessions.

        Returns:
            ProfileDiff with suggested updates.
        """
        tone_updates: list[str] = []
        structure_updates: list[str] = []

        for p in patterns:
            if p.category == "tone":
                tone_updates.append(p.description)
            elif p.category == "structure":
                structure_updates.append(p.description)

        summary_parts: list[str] = []
        if tone_updates:
            summary_parts.append(f"톤 패턴 {len(tone_updates)}건 발견")
        if structure_updates:
            summary_parts.append(f"구조 패턴 {len(structure_updates)}건 발견")

        return ProfileDiff(
            tone_updates=tone_updates,
            structure_updates=structure_updates,
            new_patterns=patterns,
            summary=". ".join(summary_parts) if summary_parts else "변경 없음",
        )

    @staticmethod
    def apply_profile_diff(
        profile_path: Path,
        diff: ProfileDiff,
    ) -> None:
        """Apply approved profile updates to profile.yaml.

        Args:
            profile_path: Path to .c2/profile.yaml.
            diff: Approved ProfileDiff.
        """
        from c4.c2.profile import load_profile, save_profile, update_learned_patterns

        profile = load_profile(profile_path)
        update_learned_patterns(
            profile,
            tone_preferences=diff.tone_updates,
            structure_preferences=diff.structure_updates,
        )
        save_profile(profile, profile_path)


# --- Helper functions ---


def _detect_tone_softening(
    deletions: list[str], additions: list[str]
) -> list[str]:
    """Detect cases where assertive language was softened."""
    assertive_markers = ["있습니다", "오류가", "잘못", "틀린", "부적절"]
    soft_markers = ["필요합니다", "생각됩니다", "확인", "바랍니다", "감사"]

    examples: list[str] = []

    for d in deletions:
        if any(m in d for m in assertive_markers):
            for a in additions:
                if any(m in a for m in soft_markers):
                    examples.append(f"'{d[:50]}' → '{a[:50]}'")
                    break

    return examples


def _detect_section_reorder(original: str, edited: str) -> bool:
    """Detect if sections were reordered."""
    import re

    orig_headers = re.findall(r"^#{1,3}\s+(.+)$", original, re.MULTILINE)
    edit_headers = re.findall(r"^#{1,3}\s+(.+)$", edited, re.MULTILINE)

    if len(orig_headers) < 2 or len(edit_headers) < 2:
        return False

    return orig_headers != edit_headers and set(orig_headers) == set(edit_headers)


def run_review_learning(
    draft_path: Path,
    final_path: Path,
    profile_path: Path | None = None,
    auto_apply: bool = False,
) -> ProfileDiff:
    """Compare review draft vs final to run persona learning.

    Reads the AI-generated draft and the user's final version, extracts
    edit patterns, and optionally applies them to profile.yaml.

    Args:
        draft_path: Path to review/.draft.md (AI original).
        final_path: Path to review/리뷰의견.md (user-edited version).
        profile_path: Path to .c2/profile.yaml. Defaults to .c2/profile.yaml.
        auto_apply: If True, apply discovered patterns to profile.yaml.

    Returns:
        ProfileDiff with discovered patterns.
    """
    if profile_path is None:
        profile_path = Path(".c2/profile.yaml")

    draft_text = draft_path.read_text(encoding="utf-8")
    final_text = final_path.read_text(encoding="utf-8")

    patterns = PersonaLearner.analyze_edits(draft_text, final_text)
    diff = PersonaLearner.suggest_profile_updates(patterns)

    if auto_apply and (diff.tone_updates or diff.structure_updates):
        PersonaLearner.apply_profile_diff(profile_path, diff)

    return diff


def run_write_learning(
    draft_path: Path,
    final_path: Path,
    profile_path: Path | None = None,
    auto_apply: bool = False,
) -> ProfileDiff:
    """Compare write draft vs final to run persona learning.

    Reads the AI-generated draft and the user's final version, extracts
    edit patterns specific to writing (voice, conciseness, terminology),
    and optionally applies them to profile.yaml.

    Args:
        draft_path: Path to write/drafts/.draft_v1.md (AI original).
        final_path: Path to user's final version.
        profile_path: Path to .c2/profile.yaml. Defaults to .c2/profile.yaml.
        auto_apply: If True, apply discovered patterns to profile.yaml.

    Returns:
        ProfileDiff with discovered patterns.
    """
    if profile_path is None:
        profile_path = Path(".c2/profile.yaml")

    draft_text = draft_path.read_text(encoding="utf-8")
    final_text = final_path.read_text(encoding="utf-8")

    patterns = PersonaLearner.analyze_edits(draft_text, final_text)

    # Tag write-domain patterns for clarity
    for p in patterns:
        if not p.description.startswith("[write]"):
            p.description = f"[write] {p.description}"

    diff = PersonaLearner.suggest_profile_updates(patterns)

    if auto_apply and (diff.tone_updates or diff.structure_updates):
        PersonaLearner.apply_profile_diff(profile_path, diff)

    return diff


def _detect_substitutions(
    deletions: list[str], additions: list[str]
) -> list[str]:
    """Detect word-level substitutions between deletions and additions."""
    examples: list[str] = []

    for d, a in zip(deletions, additions):
        if d and a and d != a:
            ratio = difflib.SequenceMatcher(None, d, a).ratio()
            if 0.4 < ratio < 0.95:
                examples.append(f"'{d[:40]}' → '{a[:40]}'")

    return examples

File: c4-core/internal/persona/persona.go
package persona

import (
	"fmt"
	"log/slog"
	"os"
	"regexp"
	"strings"
)

// EditPattern represents a pattern extracted from user edits.
type EditPattern struct {
	Category    string   `json:"category"`
	Description string   `json:"description"`
	Frequency   int      `json:"frequency"`
	Examples    []string `json:"examples"`
}

// ProfileDiff represents proposed changes to the user profile.
type ProfileDiff struct {
	ToneUpdates      []string      `json:"tone_updates"`
	StructureUpdates []string      `json:"structure_updates"`
	NewPatterns      []EditPattern `json:"new_patterns"`
	Summary          string        `json:"summary"`
}

var defaultAssertiveWords = []string{"있습니다", "오류가", "잘못", "틀린", "부적절"}
var defaultSoftWords = []string{"필요합니다", "생각됩니다", "확인", "바랍니다", "감사"}

// AnalyzeEdits extracts edit patterns from AI draft vs user final version.
func AnalyzeEdits(original, edited string) []EditPattern {
	return analyzeEdits(original, edited, defaultAssertiveWords, defaultSoftWords)
}

// analyzeEdits is the unexported helper that accepts custom tone dictionaries.
func analyzeEdits(original, edited string, assertive, soft []string) []EditPattern {
	var patterns []EditPattern

	origLines := strings.Split(original, "\n")
	editLines := strings.Split(edited, "\n")

	deletions, additions := computeLineDiff(origLines, editLines)

	// Detect tone softening
	if examples := detectToneSoftening(deletions, additions, assertive, soft); len(examples) > 0 {
		patterns = append(patterns, EditPattern{
			Category:    "tone",
			Description: "User softened tone: assertive → questioning/hedging",
			Frequency:   1,
			Examples:    truncateSlice(examples, 3),
		})
	}

	// Detect conciseness
	if len(edited) < int(float64(len(original))*0.8) && len(original) > 0 {
		ratio := float64(len(edited)) / float64(len(original))
		patterns = append(patterns, EditPattern{
			Category:    "structure",
			Description: fmt.Sprintf("User significantly shortened text (ratio: %.2f)", ratio),
			Frequency:   1,
			Examples:    []string{fmt.Sprintf("Original: %d chars → Edited: %d chars", len(original), len(edited))},
		})
	}

	// Detect section reordering
	if detectSectionReorder(original, edited) {
		patterns = append(patterns, EditPattern{
			Category:    "structure",
			Description: "User reordered sections",
			Frequency:   1,
		})
	}

	// Detect pure deletions
	if len(deletions) > 0 && len(additions) == 0 {
		patterns = append(patterns, EditPattern{
			Category:    "deletion",
			Description: "User removed content without replacement",
			Frequency:   1,
			Examples:    truncateSlice(deletions, 3),
		})
	}

	// Detect pure additions
	if len(additions) > 0 && len(deletions) == 0 {
		patterns = append(patterns, EditPattern{
			Category:    "addition",
			Description: "User added new content",
			Frequency:   1,
			Examples:    truncateSlice(additions, 3),
		})
	}

	// Detect wording substitutions
	if subs := detectSubstitutions(deletions, additions); len(subs) > 0 {
		patterns = append(patterns, EditPattern{
			Category:    "wording",
			Description: "User substituted specific wordings",
			Frequency:   1,
			Examples:    truncateSlice(subs, 3),
		})
	}

	return patterns
}

// SuggestProfileUpdates proposes profile updates based on accumulated patterns.
func SuggestProfileUpdates(patterns []EditPattern) ProfileDiff {
	var toneUpdates, structureUpdates []string

	for _, p := range patterns {
		switch p.Category {
		case "tone":
			toneUpdates = append(toneUpdates, p.Description)
		case "structure":
			structureUpdates = append(structureUpdates, p.Description)
		}
	}

	var parts []string
	if len(toneUpdates) > 0 {
		parts = append(parts, fmt.Sprintf("톤 패턴 %d건 발견", len(toneUpdates)))
	}
	if len(structureUpdates) > 0 {
		parts = append(parts, fmt.Sprintf("구조 패턴 %d건 발견", len(structureUpdates)))
	}

	summary := "변경 없음"
	if len(parts) > 0 {
		summary = strings.Join(parts, ". ")
	}

	return ProfileDiff{
		ToneUpdates:      toneUpdates,
		StructureUpdates: structureUpdates,
		NewPatterns:      patterns,
		Summary:          summary,
	}
}

// RunPersonaLearn compares draft vs final and optionally applies to profile.
func RunPersonaLearn(draftPath, finalPath, profilePath string, autoApply bool) (*ProfileDiff, error) {
	if profilePath == "" {
		profilePath = ".c2/profile.yaml"
	}

	draftBytes, err := os.ReadFile(draftPath)
	if err != nil {
		return nil, fmt.Errorf("read draft: %w", err)
	}
	finalBytes, err := os.ReadFile(finalPath)
	if err != nil {
		return nil, fmt.Errorf("read final: %w", err)
	}

	// Load profile for custom tone dictionaries; missing-file errors are silently ignored
	// (fallback to defaults), but other errors (e.g. malformed YAML) are logged.
	profile, loadErr := LoadProfile(profilePath)
	if loadErr != nil {
		slog.Warn("persona_learn: profile load failed, using default tone dicts", "error", loadErr, "path", profilePath)
	}
	if profile == nil {
		profile = map[string]any{}
	}
	assertive := profileStringSlice(profile, "learned_patterns", "tone_assertive", defaultAssertiveWords)
	soft := profileStringSlice(profile, "learned_patterns", "tone_soft", defaultSoftWords)

	patterns := analyzeEdits(string(draftBytes), string(finalBytes), assertive, soft)
	diff := SuggestProfileUpdates(patterns)

	if autoApply && (len(diff.ToneUpdates) > 0 || len(diff.StructureUpdates) > 0) {
		UpdateLearnedPatterns(profile, diff.ToneUpdates, diff.StructureUpdates, nil)
		if err := SaveProfile(profile, profilePath); err != nil {
			return nil, fmt.Errorf("save profile: %w", err)
		}
	}

	return &diff, nil
}

// profileStringSlice retrieves a string slice from profile[section][key],
// returning fallback if absent.
func profileStringSlice(profile map[string]any, section, key string, fallback []string) []string {
	if profile == nil {
		return fallback
	}
	sec, ok := profile[section].(map[string]any)
	if !ok {
		return fallback
	}
	if v := toStringSlice(sec[key]); len(v) > 0 {
		return v
	}
	return fallback
}

// =========================================================================
// Diff helpers
// =========================================================================

// computeLineDiff returns deleted and added lines using a frequency-based (bag-of-lines) diff.
func computeLineDiff(orig, edit []string) (deletions, additions []string) {
	// Build a set of lines in each
	origSet := make(map[string]int)
	editSet := make(map[string]int)
	for _, l := range orig {
		origSet[l]++
	}
	for _, l := range edit {
		editSet[l]++
	}

	// Lines in orig but not in edit = deletions
	used := make(map[string]int)
	for _, l := range orig {
		used[l]++
		if used[l] > editSet[l] {
			trimmed := strings.TrimSpace(l)
			if trimmed != "" {
				deletions = append(deletions, trimmed)
			}
		}
	}

	// Lines in edit but not in orig = additions
	used = make(map[string]int)
	for _, l := range edit {
		used[l]++
		if used[l] > origSet[l] {
			trimmed := strings.TrimSpace(l)
			if trimmed != "" {
				additions = append(additions, trimmed)
			}
		}
	}

	return
}

func detectToneSoftening(deletions, additions []string, assertive, soft []string) []string {
	var examples []string
	for _, d := range deletions {
		if containsAny(d, assertive) {
			for _, a := range additions {
				if containsAny(a, soft) {
					examples = append(examples, fmt.Sprintf("'%s' → '%s'", truncStr(d, 50), truncStr(a, 50)))
					break
				}
			}
		}
	}
	return examples
}

func detectSectionReorder(original, edited string) bool {
	headerRe := regexp.MustCompile(`(?m)^#{1,3}\s+(.+)$`)
	origHeaders := headerRe.FindAllStringSubmatch(original, -1)
	editHeaders := headerRe.FindAllStringSubmatch(edited, -1)

	if len(origHeaders) < 2 || len(editHeaders) < 2 {
		return false
	}

	origNames := make([]string, len(origHeaders))
	editNames := make([]string, len(editHeaders))
	origSet := make(map[string]bool)
	editSet := make(map[string]bool)

	for i, m := range origHeaders {
		origNames[i] = m[1]
		origSet[m[1]] = true
	}
	for i, m := range editHeaders {
		editNames[i] = m[1]
		editSet[m[1]] = true
	}

	// Same headers, different order?
	if len(origSet) != len(editSet) {
		return false
	}
	for k := range origSet {
		if !editSet[k] {
			return false
		}
	}

	// Check if order differs
	for i := range origNames {
		if i >= len(editNames) {
			return false
		}
		if origNames[i] != editNames[i] {
			return true
		}
	}
	return false
}

func detectSubstitutions(deletions, additions []string) []string {
	var examples []string
	minLen := len(deletions)
	if len(additions) < minLen {
		minLen = len(additions)
	}
	for i := 0; i < minLen; i++ {
		d, a := deletions[i], additions[i]
		if d != "" && a != "" && d != a {
			ratio := similarityRatio(d, a)
			if ratio > 0.4 && ratio < 0.95 {
				examples = append(examples, fmt.Sprintf("'%s' → '%s'", truncStr(d, 40), truncStr(a, 40)))
			}
		}
	}
	return examples
}

// similarityRatio computes a simple character overlap ratio.
func similarityRatio(a, b string) float64 {
	if len(a) == 0 && len(b) == 0 {
		return 1.0
	}
	// Count matching characters using a simple approach
	shorter, longer := a, b
	if len(a) > len(b) {
		shorter, longer = b, a
	}
	matches := 0
	used := make([]bool, len(longer))
	for _, c := range shorter {
		for j, lc := range longer {
			if !used[j] && c == lc {
				matches++
				used[j] = true
				break
			}
		}
	}
	return 2.0 * float64(matches) / float64(len(a)+len(b))
}

func containsAny(s string, markers []string) bool {
	for _, m := range markers {
		if strings.Contains(s, m) {
			return true
		}
	}
	return false
}

func truncStr(s string, maxLen int) string {
	r := []rune(s)
	if len(r) > maxLen {
		return string(r[:maxLen])
	}
	return s
}

func truncateSlice(s []string, max int) []string {
	if len(s) > max {
		return s[:max]
	}
	return s
}
